### DISCUSSION TITLE
Q:--max-streams-bidi=<N> is set to 0 or absent

### DISCUSSION THREAD

**Discussion by devkdas** [ngtcp2/ngtcp2#1031] [Q&A]
@tatsuhiro-t built qtlsclient and qtlsserver using ngtcp2 and trying to run the client and server in my ubuntu machines

while going through options i have found out that you have set this value as 0 and in pcap this value is totally absent
```
  --max-streams-bidi=<N>
              The number of the concurrent bidirectional streams.
              Default: 0
```
File download is fine but as per RFC9000 

### initial_max_streams_bidi (0x08):
The initial maximum bidirectional streams parameter is an integer value that contains the initial maximum number of bidirectional streams the endpoint that receives this transport parameter is permitted to initiate. **If this parameter is absent or zero, the peer cannot open bidirectional streams until a MAX_STREAMS frame is sent.** Setting this parameter is equivalent to sending a MAX_STREAMS ([Section 19.11](https://datatracker.ietf.org/doc/html/rfc9000#frame-max-streams)) of the corresponding type with the same value.

am i missing anything here

**Comment by tatsuhiro-t** [comment]
I am not sure what you are asking, but in HTTP/3 server is not allowed to create bidirectional streams.  Client is not required to send nonzero initial_max_streams_bidi.  I think this option was added for testing purpose.

**Comment by devkdas** [comment]
as per RFC9000 If this "**initial_max_streams_bidi**" is absent or zero, the peer cannot open _bidirectional_ streams.so client can't download file and client wont download file in _unidirectional_ stream.
but in your case even if we set "**initial_max_streams_bidi**" to **zero** on client side (**which is default value**) file is still downloading fine.
```
 --max-streams-bidi=<N>
              The number of the concurrent bidirectional streams.
              Default: 0
 ```

so i was in confusion. am i missing anything here or did we implement client in wrong way   

RFC9000 ref:-   https://datatracker.ietf.org/doc/html/rfc9000#section-18.2-4.22.1

**Comment by tatsuhiro-t** [comment]
You are certainly confused but don't worry this is one of the most complicated part of this specification that needs careful understandings of how transport parameters work.  Even implementers are sometimes confused by this.

The number of streams that client can open is governed by server transport parameters.
What you are referring to is client transport parameters, which governs the stream creation by server.

qtlsclient is HTTP/3 client.  HTTP/3 does not allow server to open bidirectional streams.  Therefore, it is completely fine for client to send initial_max_streams_bidi to 0 because server is not supposed to open bidirectional streams.

qtlsclient can open stream and do request and get response fine because server sends initial_max_streams_bidi > 0.

I hope this clarifies your concerns.

**Comment by devkdas** [comment]
Thanks @tatsuhiro-t
